staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go

```diff
@@ -14,76 +14,43 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// This file provides abstractions for setting the provider (e.g., prometheus)
-// of metrics.
-
 package cache
 
 import (
-	"sync"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
-// GaugeMetric represents a single numerical value that can arbitrarily go up
-// and down.
-type GaugeMetric interface {
-	Set(float64)
-}
-
-// CounterMetric represents a single numerical value that only ever
-// goes up.
-type CounterMetric interface {
-	Inc()
-}
-
-// SummaryMetric captures individual observations.
-type SummaryMetric interface {
-	Observe(float64)
-}
-
-type noopMetric struct{}
-
-func (noopMetric) Inc()            {}
-func (noopMetric) Dec()            {}
-func (noopMetric) Observe(float64) {}
-func (noopMetric) Set(float64)     {}
-
-// MetricsProvider generates various metrics used by the reflector.
-type MetricsProvider interface {
-	NewListsMetric(name string) CounterMetric
-	NewListDurationMetric(name string) SummaryMetric
-	NewItemsInListMetric(name string) SummaryMetric
-
-	NewWatchesMetric(name string) CounterMetric
-	NewShortWatchesMetric(name string) CounterMetric
-	NewWatchDurationMetric(name string) SummaryMetric
-	NewItemsInWatchMetric(name string) SummaryMetric
-
-	NewLastResourceVersionMetric(name string) GaugeMetric
-}
-
-type noopMetricsProvider struct{}
+const (
+	reflectorSubsystem = "reflector"
+)
 
-func (noopMetricsProvider) NewListsMetric(name string) CounterMetric         { return noopMetric{} }
-func (noopMetricsProvider) NewListDurationMetric(name string) SummaryMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewItemsInListMetric(name string) SummaryMetric   { return noopMetric{} }
-func (noopMetricsProvider) NewWatchesMetric(name string) CounterMetric       { return noopMetric{} }
-func (noopMetricsProvider) NewShortWatchesMetric(name string) CounterMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewWatchDurationMetric(name string) SummaryMetric { return noopMetric{} }
-func (noopMetricsProvider) NewItemsInWatchMetric(name string) SummaryMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewLastResourceVersionMetric(name string) GaugeMetric {
-	return noopMetric{}
-}
+var (
+	// reflectorListCount is the number of list requests issued by a reflector in order to start
+	// a watch broken down by the group-version-kind and the resource-version used.
+	reflectorListCount = prometheus.NewCounterVec(
```

Comment:
The problem with using component-base is that it introduces a circular dependency: k8s.io/component-base -> k8s.io/client-go -> k8s.io/component-base 
Thoughts?

Reply:
It's weird. Let me have a look.

Reply:
I haven't found an import cycle. Can you share the error message? 

My dummy code :
```diff
diff --git a/staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go b/staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go
index 5c00115f59b..f667fe1cd56 100644
--- a/staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go
+++ b/staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go
@@ -21,8 +21,21 @@ package cache

 import (
        "sync"
+
+       "k8s.io/component-base/metrics"
+       "k8s.io/component-base/metrics/legacyregistry"
 )

+var reflectorListCount = metrics.NewCounterVec(
+       &metrics.CounterOpts{
+               Name: "watch_list_count",
+       },
+       []string{"resource_type", "resource_version"})
+
+func init() {
+       legacyregistry.MustRegister(reflectorListCount)
+}
+
 // GaugeMetric represents a single numerical value that can arbitrarily go up
 // and down.
 type GaugeMetric interface {

```

Reply:
The cycle is at the go.mod level.

If I want to use `k8s.io/component-base` in `k8s.io/client-go/tools/cache/`, I need to add `k8s.io/component-base` to the [k8s.io/client-go/go.mod](https://github.com/kubernetes/client-go/blob/master/go.mod) file. This will introduce a circular dependency as [k8s.io/component-base/go.mod lists k8s.io/client-go as a dependency](https://github.com/kubernetes/component-base/blob/d868452632e2f3b000751e0050777d0f8c43e69e/go.mod#L17).

Reply:
Yeah, I see it. Thanks. Let me think about how to deal with this.

Reply:
We purposefully do not define or declare metrics directly in client-go for a number of reasons, this being one of them. 

The standard pattern is to use interfaces for your metrics and use an adapter for dependency injection. One such example is [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/leaderelection/healthzadaptor.go). 

You'll probably want to do the same. The actual implementation of metrics can be placed into `component-base/metrics/prometheus/clientgo/`.

Reply:
Thanks Han, I'll take a look at the examples you mentioned and try to this it this way (probably in early Jan).